doc/normbench.1
'\" -*- nroff -*-
'\" Copyright (c) 2001 by Andreas Kupries <a.kupries@westend.com>
'\" All rights reserved.
'\" 
'\" RCS: @(#) $Id: normbench.1,v 1.1 2001/05/29 21:16:16 andreas_kupries Exp $
'\" 
.so man.macros
.TH normbench.tcl 1 1.0 Normbench "Tclbench application"
.BS
'\" Note:  do not modify the .SH NAME line immediately below!
.SH NAME
normbench.tcl \- Normalize benchmark results.
.SH SYNOPSIS
\fBnormbench.tcl\fR \fI?-option value ...? ?file?\fR
.BE
.SH DESCRIPTION
.PP
The \fBnormbench.tcl\fR application provides the same functionality as
the \fI-normalize\fR option of \fBrunbench.tcl\fR. This allows the
postprocessing of benchmark results which there not normalized right
away. It processes either the file whose name was specified on the
commandline or \fBstdin\fR. It automatically detects which of the
formats supported by \fBrunbench.tcl\fR the input is in.
.PP
The following options are recognized by the application:
.TP
\fB-help\fR
Causes the application to print a list of the recognized option plus
short explanations of their meaning.
.TP
\fB-normalize\fR
Takes a version number as value and causes the application to
normalize the timing values to the value of the given version.
.RE
.SH KEYWORDS
benchmarks, tclbench, libbench, runbench


ChangeLog
2001-05-29  Andreas Kupries  <andreas_kupries@users.sourceforge.net>

	* doc/libbench.n:
	* doc/runbench.1:
	* doc/normbench.1: Added documentation of benchmark library and
	  applications.

	* doc: Added documentation directory.

2001-05-22  Jeff Hobbs  <jeffh@ActiveState.com>

	* runbench.tcl: corrected error for reporting errors in sourced files

	* tcl/fcopy.bench: made use of bench_tmpfile for more accurate
	data (not skewed by network).

	* libbench.tcl (bench_tmpfile): correctly allow multiple calls to
	bench_tmpfile within one file.

	* normbench.tcl: new file that allows for post-process
	normalization of the benchmark data.
	Corrected last minute code checkin bug.
	Added support for moving left (to higher versions) to normalize
	when the requested version returned non-double data.

	* tcl/libbench.tcl:
	* tcl/runbench.tcl: changed -iterations to be a maximum number for
	timings, to override any larger number the benchmark may set for
	itself.
	Rearranged result format of benchmarks to return data by benchmark
	description.  Benchmarks are now always returned in alphabetical
	order of the benchmark description.
	Changed benchmarks to rerun the interpreter per benchmark file
	instead of sourcing all files into the same interpreter.  This
	reduces any skew related to excessive mem usage or other factors
	that may arise for one benchmark file.
	Changed midpoint numbers to time elapsed calculation.
	Added -normalize option that post-processes the time information
	to normalize against one version as a baseline.
	Changed -errors <bool> to -throwerrors with no arg, and changed
	the default to not throw errors in benchmark files.
	Added version string to verbose run info.

	* tcl/klist.bench: added support for <8.0 to all benchmarks except
	shuffle0, with notably reduced default run iters due to extreme
	slowness of <8.0 interps for these tasks.

	* tcl/string.bench:
	* tcl/regexp.bench: fixed incorrect str-repeat replacement function

2001-05-18  Jeff Hobbs  <jeffh@ActiveState.com>

	* tcl/string.bench: added <8.0 compatible rev-recursive benchmark,
	fixed non-octal escape in ustring instantiation.

	* tcl/wordcount.bench: added <8.1 compatible benchmarks

	* tcl/methods.bench: return for interps <8.0

2001-05-19  Andreas Kupries  <andreas_kupries@users.sourceforge.net>

	* tcl/conditional.bench: Changed some descriptions to make them
	  unique and matching to the code.

	* tcl/fcopy.bench: New benchmarks for the [fcopy] command
	  (unsupported0 in older versions of the core).

2001-05-16  Jeff Hobbs  <jeffh@ActiveState.com>

	* tcl/string.bench: added static string length benchmarks

	* tcl/wordcount.in: 
	* tcl/wordcount.bench: wordcount benchmarks

	* tcl/heapsort.bench: new file with heapsort benchmark
	* tcl/string.bench:
	* tcl/matrix.bench:
	* tcl/regexp.bench: extended benchmarks

2001-05-11  Jeff Hobbs  <jeffh@ActiveState.com>

	* tcl/string.bench: clarified string reverse benchmarks, added
	more to the string compare benchmarks.

	* tcl/matrix.bench: some new matrix benchmarks.  Basically a seed
	file looking for more.  procs courtesy Sofer.

	* tcl/list.bench: added a list-iter benchmark

	* tcl/klist.bench: reduced default iters in klist.bench.  Accuracy
	seems about the same without the wait...

	* libbench.tcl: 
	* runbench.tcl: added support for -rmatch option (regexp match of
	benchmark description).
	Added MIDPOINT verbose puts for interim time notes.

2001-04-11  Jeff Hobbs  <jeffh@ActiveState.com>

	* tcl/klist.bench: added shuffle5* from wiki.

2001-03-28  Jeff Hobbs  <jeffh@activestate.com>

	* tcl/string.bench: fixed str-first proc that had bogus code in it.
	added more split benchmarks for dkf's split improvement in 8.4.
	
	* tk/canvas.bench: expanded item draw benchmarks

2001-03-23    <jeffh@activestate.com>

	* tk/canvas.bench: added simple item draw benchmarks

2001-03-15    <jeffh@activestate.com>

	* tcl/klist.bench: improved non-tclbench data output.

	* runbench.tcl: added more error capturing.

	* tcl/string.bench: fixed calls to string repeat to work with
	<8.1.1 interps.

	* tcl/klist.bench: new file to benchmark various list shuffling
	techniques (from wiki).
	* tcl/methods.bench: new file to benchmark various method
	invocation speeds (petasis).

2000-10-19  Jeff Hobbs  <hobbs@ajubasolutions.com>

	* tcl/string.bench (str-append-2): added more append tests

2000-08-30  Jeff Hobbs  <hobbs@scriptics.com>

	* tcl/string.bench: made string repeat calls compatible with
	pre-8.1.1 interpreters.

	* libbench.tcl (bench_tmpfile): add env to global list

2000-08-29  Eric Melski  <ericm@ajubasolutions.com>

	* tcl/string.bench: Extended string append benchmarks to exploit
	new growth algorithm for string objects in Tcl 8.4a2.

2000-05-31  Jeff Hobbs  <hobbs@scriptics.com>

	* runbench.tcl: new options -errors (passed to libbench), -verbose
	(by default we are now quieter on output), -output <text|list|csv>
	(different output types - csv is char-sep-value for Excel).
	Added start/finish times (in -verbose mode).
	* libbench.tcl: libbench now takes -option switches for
	flexibility, options for -errors BOOL (error suppression), -interp
	NAME (to specify interp), -match PATTERN (glob pattern to filter
	tests by desc), -iters NUM (default number of iters to run).
	Reorganized how data is returned to runbench master.

	* tk/entry.bench (new):
	* tk/canvas.bench (new): new tests for widget creation, config

	* tcl/array.bench (removed):
	* tcl/vars.bench: merged array.bench tests into VAR

	* tcl/map.bench: fixed for compatability with Tcl7.4-

2000-05-25  Jeff Hobbs  <hobbs@scriptics.com>

	* runbench.tcl: added -match, -notcl, -notk options, restructured
	startup sequence.

	* libbench.tcl: added ability to return string values from bench
	tests and support for filtering tests to run.

	* tcl/string.bench: moved string mapping benchmarks and added more
	string equality benchmarks
	* tcl/map.bench: added extended string mapping benchmark

	* tcl/read.bench:

normbench.tcl
#!/bin/sh
# The next line is executed by /bin/sh, but not tcl \
exec tclsh "$0" ${1+"$@"}

# runbench.tcl ?options?
#
set RCS {RCS: @(#) $Id: normbench.tcl,v 1.2 2001/05/23 04:06:08 hobbs Exp $}
#
# Copyright (c) 2000-2001 Jeffrey Hobbs.

#
# Run the main script from an 8.2+ interp
#
if {[catch {package require Tcl 8.2}]} {
    set me [file tail [info script]]
    puts stderr "$me requires 8.2+ to run, although it can benchmark\
	    any Tcl v7+ interpreter"
    exit 1
}

regexp {,v (\d+\.\d+)} $RCS -> VERSION
set MYDIR [file dirname [info script]]
set ME [file tail [info script]]

proc usage {} {
    puts stderr "Usage (v$::VERSION): $::ME ?options?\
	    \n\t-help			# print out this message\
	    \n\t-normalize <version>	# normalize numbers to given version\
	    \n\t?file?			# runbench output file to normalize (or stdin)"
    exit 1
}

#
# Process args
#
array set opts {
    norm	{}
    fid		stdin
}
if {[llength $argv]} {
    while {[llength $argv]} {
	set key [lindex $argv 0]
	switch -glob -- $key {
	    -help*	{ usage }
	    -norm*	{
		set opts(norm) [lindex $argv 1]
		set argv [lreplace $argv 0 1]
	    }
	    default {
		if {![file exists $key]} {
		    usage
		} else {
		    set opts(fid) [open $key r]
		    set argv [lreplace $argv 0 0]
		    # The file should be the last arg
		    if {[llength $argv]} { usage }
		}
	    }
	}
    }
}

proc csv2list {str {sepChar ,}} {
    regsub -all {(\A\"|\"\Z)} $str \0 str
    set str [string map [list $sepChar\"\"\" $sepChar\0\" \
	    \"\"\"$sepChar \"\0$sepChar \
	    \"\" \" \" \0 ] $str]
    set end 0
    while {[regexp -indices -start $end {(\0)[^\0]*(\0)} $str \
	    -> start end]} {
	set start [lindex $start 0]
	set end   [lindex $end 0]
	set range [string range $str $start $end]
	set first [string first $sepChar $range]
	if {$first >= 0} {
	    set str [string replace $str $start $end \
		    [string map [list $sepChar \1] $range]]
	}
	incr end
    }
    set str [string map [list $sepChar \0 \1 $sepChar \0 {} ] $str]
    return [split $str \0]
}

proc list2csv {list {sepChar ,}} {
    set out ""
    foreach l $list {
	set sep {}
	foreach val $l {
	    if {[string match "*\[\"$sepChar\]*" $val]} {
		append out $sep\"[string map [list \" \"\"] $val]\"
	    } else {
		append out $sep$val
	    }
	    set sep $sepChar
	}
	append out \n
    }
    return $out
}


proc findVersion {norm versions} {
    if {$norm == ""} { return 0 }
    set i [lsearch -exact $versions $norm]
    if {$i >= 0} { return $i }
    set i [lsearch -glob $versions *$norm*]
    if {$i >= 0} { return $i }
    puts stderr "Unable to normalize \"$norm\": must be one of [join $versions {, }]"
    usage
}

proc normalize-text {norm line} {
    global start col
    scan $line %d num
    if {$num == 0} {
	set start [expr {[string first 1: $line]-1}]
	set col [findVersion $norm [string range $line $start end]]
	return $line
    }
    set times [string range $line $start end]
    set ntime [lindex $times $col]
    if {![string is double -strict $ntime]} {
	# This didn't return valid data.  Try walking backwards to find
	# a newer version that we can normalize this row on, since newer
	# versions are to the left.
	for {set i $col} {$i >= 0} {incr i -1} {
	    set ntime [lindex $times $i]
	    if {[string is double -strict $ntime]} { break }
	}
	# Hmph.  No usable data.
	if {$i == -1} { return $line }
    }
    set out [string range $line 0 [expr {$start-1}]]
    foreach t $times {
	if {[string is double -strict $t]} {
	    append out [format " %7.2f" \
		    [expr {double($t) / double($ntime)}]]
	} else {
	    append out [format " %7s" $t]
	}
    }
    return $out 
}

proc normalize-list {norm line} {
    global col
    if {[lindex $line 0] == 0} {
	set col [findVersion $norm [lrange $line 2 end]]
	return $line
    }
    set times [lrange $line 2 end]
    set ntime [lindex $times $col]
    if {![string is double -strict $ntime]} {
	return $line
    } else {
	set out [lrange $line 0 1]
	foreach t $times {
	    if {[string is double -strict $t]} {
		lappend out [format "%.2f" \
			[expr {double($t) / double($ntime)}]]
	    } else {
		lappend out $t
	    }
	}
	return $out 
    }
}

proc normalize {norm indata} {
    set lines [split $indata \n]
    foreach line $lines {
	if {![string match "\[0-9\]*" $line]} {
	    puts stdout $line
	    continue
	}
	scan $line %d num
	if {$num == 0} {
	    # guess format based on first line of version input
	    if {[string match "0,VER*" $line]} {
		set format csv
	    } elseif {[string match "0 VER*" $line]} {
		set format list
	    } elseif {[string match "000*VER*" $line]} {
		set format text
	    } else {
		puts stderr "Unrecognized runbench format input file"
		exit
	    }
	}
	switch -exact -- $format {
	    text { puts stdout [normalize-text $norm $line] }
	    list { puts stdout [normalize-list $norm $line] }
	    csv  {
		puts -nonewline stdout [list2csv [list \
			[normalize-list $norm [csv2list $line]]]]
	    }
	}
    }
}

normalize $opts(norm) [read -nonewline $opts(fid)]
